Add documentation on how to configure user-specific XKB RMLVO (#150)
xkbcommon/libxkbcommon: doc/user-configuration.md

# User-configuration

This page describes how to add a custom layout or option so that it will be
parsed by libxkbcommon.

**The below requires libxkbcommon as keymap compiler and does not work in X**.

## Data locations

libxkbcommon searches the following paths for XKB configuration files:

> - `$XDG_CONFIG_HOME/xkb/`
> - `$HOME/.config/xkb/`

Reviewer: ```suggestion
- `$XDG_CONFIG_HOME/xkb/`, or `$HOME/.config/xkb/` if the `XDG_CONFIG_HOME` environment variable is not defined
```

- `$HOME/.xkb/`
- `$XKB_CONFIG_ROOT`

> - `<prefix>/X11/xkb/` (on most distributions this is `/usr/share/X11/xkb`)

Reviewer: ```suggestion
- `<datadir>/X11/xkb/` (path defined by the xkeyboard-config package, on most distributions this is `/usr/share/X11/xkb`)
```

A keymap created with `xkb_keymap_new_from_names()` will look up those paths in
order until the required data is found.

> **Note: Where libxkbcommon runs in a privileged context, only the prefix path is

Reviewer: ```suggestion
**Note: Where libxkbcommon runs in a privileged context, only the system (datadir) path is
```

- `keycodes`
- `rules`
- `symbols`
- `types`

## RMLVO vs KcCGST

Due to how XKB is configured, there is no such thing as a "layout" in XKB
itself, or, indeed, any of the rules, models, variant, options (RMLVO) decribed
in `struct xkb_rule_names`. RMLVO names are merely lookup keys in the
rules file provided by xkeyboard-config to map to the correct keycode, compat,
geometry (ignored by libxkbcommon), symbols and types (KcCGST). The KcCGST data
is the one used by XKB and libxbkcommon to map keys to actual symbols.

For example, a common RMLVO configuration is layout "us", variant "dvorak" and
option "terminate:ctrl_alt_bksp". Using the default rules file and model
this maps into the following KcCGST components:

```
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+us(dvorak)+inet(evdev)+terminate(ctrl_alt_bksp)"	};
	xkb_geometry  { include "pc(pc105)"	};
};
```

A detailed explanation of how rules files convert RMLVO to KcCGST is out of
scope for this document. See [the rules file](md_doc_rules-format.html) page
instead.

## Adding a layout

> Adding a layout requires that the user adds **symbols** in the correct location.

Reviewer: ```suggestion
Adding a layout requires that the user add **symbols** in the correct location.
```

Author: that doesn't seem right. "the user adds" or "users add". Given that it's user-specific, I'd prefer to stick to singular here.

> The default rules files (usually `evdev`) have a catch-all to map a layout "foo"
> variant "bar" into the "bar" section in the file `$xkb_base_dir/symbols/foo`.

Reviewer: ```suggestion
The default rules files (usually `evdev`) have a catch-all to map a layout, say "foo",
and a variant, say "bar", into the "bar" section in the file `symbols/foo`.
```

This is sufficient to define a new keyboard layout. The example below defines
the keyboard layout "banana" with an optional variant "orange"

```
$ cat $XDG_CONFIG_HOME/xkb/symbols/banana

> // Like a US layout but swap the top row's so numbers are on Shift

Reviewer: ```suggestion
// Like a US layout but swap the top row so numbers are on Shift
```

default partial alphanumeric_keys
xkb_symbols "basic" {
    include "us(basic)"
    name[Group1]= "Banana (US)";

>     key <AE01> {	[	exclam, 	1]	};

Reviewer: I think it will be better to replace the tabs with spaces, their rendering is rather inconsistent.

    key <AE02> {	[	at, 		2]	};
    key <AE03> {	[	numbersign, 	3]	};
    key <AE04> {	[	dollar, 	4]	};
    key <AE05> {	[	percent, 	5]	};
    key <AE06> {	[	asciicircum, 	6]	};
    key <AE07> {	[	ampersand, 	7]	};
    key <AE08> {	[	asterisk, 	8]	};
    key <AE09> {	[	parenleft, 	9]	};
    key <AE10> {	[	parenright, 	0]	};
    key <AE11> {	[	underscore, 	minus]	};
    key <AE12> {	[	plus, 	equal]	};
};

// Same as banana but map the euro sign to the 5 key
partial alphanumeric_keys
xkb_symbols "orange" {
    include "banana(basic)"
    name[Group1] = "Banana (Eurosign on 5)";
    include "eurosign(5)"
};
```

The `default` section is loaded when no variant is given. The first example
sections uses ``include`` to populate with a symbols list defined elsewhere
(here: section `basic` from the file `symbols/us`, aka. the default US keyboard
layout) and overrides parts of these
symbols. The effect of this section is to swap the numbers and symbols in the

> top-most row (compared to the US layout) but otherwise uses the US layout.

Reviewer: ```suggestion
top-most row (compared to the US layout) but otherwise use the US layout.
```
